libraries/libfc/include/fc/network/listener.hpp

```diff
@@ -0,0 +1,211 @@
+#pragma once
+#include <boost/asio/deadline_timer.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/ip/v6_only.hpp>
+#include <boost/asio/local/stream_protocol.hpp>
+#include <filesystem>
+#ifdef __cpp_concepts
+#include <concepts>
+#endif
+
+#include <fc/log/logger.hpp>
+#include <fc/scoped_exit.hpp>
+
+namespace fc {
+
+inline std::string to_string(const boost::asio::ip::tcp::endpoint& endpoint) {
+   const auto& ip_addr        = endpoint.address();
+   std::string ip_addr_string = ip_addr.to_string();
+   if (ip_addr.is_v6()) {
+      ip_addr_string = "[" + ip_addr_string + "]";
+   }
+   return ip_addr_string + ":" + std::to_string(endpoint.port());
+}
+
+inline std::pair<std::string, std::string> split_host_port(std::string_view endpoint) {
+   std::string::size_type colon_pos = endpoint.rfind(':');
+   if (colon_pos != std::string::npos) {
+      auto port = endpoint.substr(colon_pos + 1);
+      auto hostname =
+            (endpoint[0] == '[' && colon_pos >= 2) ? endpoint.substr(1, colon_pos - 2) : endpoint.substr(0, colon_pos);
+      return { std::string(hostname), std::string(port) };
+   } else {
+      return { std::string(endpoint), {} };
+   }
+}
+
+#ifdef __cpp_concepts
+template <class T>
+concept listener_state = requires(T state) {
+   { state->thread_pool.get_executor() } -> std::same_as<boost::asio::io_context&>;
+   { state->get_logger() } -> std::same_as<fc::logger&>;
+};
+#endif
+
+template <typename T, typename Protocol, typename State>
+#ifdef __cpp_concepts
+   requires listener_state<State>
+#endif
+struct listener : std::enable_shared_from_this<T> {
+   using endpoint_type = typename Protocol::endpoint;
+
+   State                       state_;
+   typename Protocol::acceptor acceptor_;
+   boost::asio::deadline_timer accept_error_timer_;
+
+   listener(State state, const endpoint_type& endpoint)
+       : state_(state), acceptor_(state->thread_pool.get_executor(), endpoint),
+         accept_error_timer_(state->thread_pool.get_executor()) {}
+
+   void do_accept() {
+      acceptor_.async_accept(state_->thread_pool.get_executor(),
+                             [self = this->shared_from_this()](boost::system::error_code ec, auto&& peer) {
+                                self->on_accept(ec, std::forward<decltype(peer)>(peer));
+                             });
+   }
+
+   template <typename Socket>
+   void on_accept(boost::system::error_code ec, Socket&& socket) {
+      if (!ec) {
+         static_cast<T*>(this)->create_session(std::forward<Socket>(socket));
+         do_accept();
+      } else if (ec == boost::system::errc::too_many_files_open) {
+         // retry accept() after timeout to avoid cpu loop on accept
+         auto timeout_ms = static_cast<T*>(this)->accept_timeout_ms;
```

Comment:
I this this could either be an argument to the constructor or just normalize everyone to 100ms. My vote would be to just make it a const 100ms for every listener.